added cluster manager for mpi

amitmurthy opened the pull request:
Allows for an mpi cluster to be started from the julia master. Hence makes it possible for julia parallel methods to be used along with code making MPI calls.

amitmurthy commented:
cc @lcw @andreasnoack @eschnett @ViralBShah

ViralBShah commented:
Exciting. Looking forward to trying when merged.

ViralBShah commented:
Can I use multiple cluster managers. Often, you may use one cluster manager to allocate CPUs, and then want to run an MPI job on those CPUs.

amitmurthy commented:
Yes, you should be able to. The macro `@mpi_do` will need to support it to make it easier for folks  though. Currently  `@mpi_do` executes on all workers() and not just a subset of them.  But as currently implemented `manager.map_julia_mpi` has a list of all pids that are part of this MPI cluster and you can do a SPMD style parallel execution on only those workers.

lcw commented:
This looks like a nice addition to MPI.jl.  Just a couple of questions.  What are you calling making sure ` myid() == 1` before you call `MPI.Init()` and `MPI.Finalize()`?  Should those check be put into `MPI.Init()` and `MPI.Finalize()`?

Also do you know if `atexit()` is called after the garbage collector cleans up all the memory?  I had problems before calling MPI_Finalize() before all of the MPI objects were cleaned up.

andreasnoack commented:
I have now tried your pr and it looks very promising. I'm able to call ScaLAPACK's SVD from a julia prompt now. That's great. A couple of observations:

- I cannot run it on 0.3. That might be okay as 0.4 is approaching, but since you are requiring `compat` i thought it should work.
- It appears that `addprocs(MPICluster)` call whatever `julia` there is in the path. I use a symlink to `julia-dev` for `0.4` and `julia` for `0.3` so when starting the cluster from `0.4` I get `0.3` workers. This is different from the behavior of socket processes.
- I received some errors that terminated the processes, but it seemed that they were still running as MPI workers, i.e. they were not present in `procs`, but when calling `@mpi_do println("Hello!")`, they still responded.

amitmurthy commented:
@lcw `05-juliacman.jl` is meant to be executed directly - `julia 05-juliacman.jl` and not under `mpirun`. Since `05-juliacman.jl` executes all the other examples - viz., `01-hello.jl,  02-broadcast.jl  03-reduce.jl and  04-sendrecv.jl `, they must not execute MPI_Init and MPI_Finalize when run in this manner. When 01-hello.jl is directly run under `mpirun` all the mpi processes will have julia pid of 1 and they need to run init/finalize.  Under the MPIManager, `mpi_init` is called to execute init/finalize on all workers.

I am not sure of the behavior of `atexit` w.r.t. `gc`, but so far in my testing it has not been a problem.

amitmurthy commented:
@andreasnoack `compat` was required to remove the bunch of warnings thrown up by the existing MPI.jl codebase. In hindsight it is not relevant, since the 0.4 clustermanager interface is different from the 0.3 one, so it makes since to just add a `julia 0.4` tag to REQUIRE and drop `compat` altogether.

The launch command by default is `mpirun -np $np --output-filename $output_filename julia --worker`. You can override this in the MPIManager constructor via keyword arg `launch_cmd`. The path may also be different when the MPI workers are on a bunch of different nodes. But, yes, the default should at least pick the same julia version as the master.

Are you sure about the error case? `@mpi_do` executes on all workers only - in which case they must be present in `procs` and the workers are not terminated. The MPI calls may have failed, but the processes must have been up and still available as julia workers.

andreasnoack commented:
Consider the following example where where it is reported that the workers terminate when I try to initialize MPI twice.
```julia
julia> using MPI

julia> manager=MPIManager(np=4)
MPI.MPIManager(4,`mpirun -np 4 --output-filename /tmp/user/1021/juliafYGoOa julia --worker`,"/tmp/user/1021/juliafYGoOa",60,Dict{Int64,Int64}(),Dict{Int64,Int64}(),PipeServer(active),"/tmp/user/1021/juliauQHs4q")

julia> addprocs(manager)--------------------------------------------------------------------------
mpirun noticed that the job aborted, but has no info as to the process
that caused that situation.
--------------------------------------------------------------------------
julia> addprocs(manager)
4-element Array{Any,1}:
 2
 3
 4
 5

julia> @everywhere using MPI

julia> @mpi_do println("Hello!")
	From worker 2:	Hello!
	From worker 3:	Hello!
	From worker 4:	Hello!
	From worker 5:	Hello!

julia> mpi_init(manager)

julia> mpi_init(manager)
[julia:48633] *** An error occurred in MPI_Init
[julia:48633] *** on communicator MPI_COMM_WORLD
[julia:48633] *** MPI_ERR_OTHER: known error not in list
[julia:48633] *** MPI_ERRORS_ARE_FATAL: your MPI job will now abort
--------------------------------------------------------------------------
Calling MPI_Init or MPI_Init_thread twice is erroneous.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun has exited due to process rank 0 with PID 48630 on
node julia exiting improperly. There are two reasons this could occur:

1. this process did not call "init" before exiting, but others in
the job did. This can cause a job to hang indefinitely while it waits
for all processes to call "init". By rule, if one process calls "init",
then ALL processes must call "init" prior to termination.

2. this process called "init", but exited without calling "finalize".
By rule, all processes that call "init" MUST call "finalize" prior to
exiting or it will be considered an "abnormal termination"

This may have caused other processes in the application to be
terminated by signals sent by mpirun (as reported here).
--------------------------------------------------------------------------
[julia:48629] 3 more processes have sent help message help-mpi-errors.txt / mpi_errors_are_fatal
[julia:48629] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
Worker 2 terminated.ERROR: ProcessExitedException()
 in wait at ./task.jl:288
 in wait at ./task.jl:198
 in wait_full at ./multi.jl:586
 in remotecall_fetch at multi.jl:686
 in call_on_owner at ./multi.jl:733
 in fetch at multi.jl:741
 in mpi_init at /home/andreasnoackjensen/.julia/v0.4/MPI/src/cman.jl:92

Worker 3 terminated.Worker 4 terminated.Worker 5 terminated.



julia> @mpi_do println("Hello!")
Hello!

julia> procs()
1-element Array{Int64,1}:
 1

julia> addprocs(manager)
Error starting cluster workers : connect: connection refused (ECONNREFUSED)
tail: tail: /tmp/user/1021/juliafYGoOa.1.0: file truncated
/tmp/user/1021/juliafYGoOa.1.0: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.1: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.2: file truncatedtail: 
/tmp/user/1021/juliafYGoOa.1.2: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.3: file truncated
	From worker 2:	julia_worker:9009#18.4.43.32
	From worker 4:	julia_worker:9010#18.4.43.32
0-element Array{Any,1}

julia> procs()
1-element Array{Int64,1}:
 1

julia> addprocs(manager)
tail: /tmp/user/1021/juliafYGoOa.1.0: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.0: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.0: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.1: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.1: file truncated
tail: tail: /tmp/user/1021/juliafYGoOa.1.2: file truncated
/tmp/user/1021/juliafYGoOa.1.2: file truncated
tail: tail: /tmp/user/1021/juliafYGoOa.1.3: file truncated
tail: /tmp/user/1021/juliafYGoOa.1.3: file truncated
/tmp/user/1021/juliafYGoOa.1.2: file truncated
	From worker 2:	julia_worker:9013#18.4.43.32
	From worker 7:	julia_worker:9013#18.4.43.32
	From worker 8:	julia_worker:9014#18.4.43.32
	From worker 9:	julia_worker:9015#18.4.43.32
	From worker 4:	julia_worker:9015#18.4.43.32
	From worker 10:	julia_worker:9016#18.4.43.32
4-element Array{Any,1}:
  7
  8
  9
 10

julia> @everywhere using MPI

julia> @mpi_do println("Hello!")
	From worker 10:	Hello!
	From worker 8:	Hello!
	From worker 2:	Hello!
	From worker 7:	Hello!
	From worker 9:	Hello!
	From worker 4:	Hello!
```
First time I try to add processes after the crash no processes are added. The second time processes are added, but it seems that it also captures some of the original processes that were reported dead.

Another thing: Would it be possible to detect if MPI has already been initialized and give a Julia error instead?

eschnett commented:
You can call `MPI_Isinitialized` to check whether MPI has already been initialized. In this case, you can either abort with an error (but why?), or continue without calling `MPI_Init`.

andreasnoack commented:
Yes. Continuing without calling `MPI_Init` makes more sense. I just wanted to avoid the worker terminations.

eschnett commented:
I just realize the function is called `MPI_Initialized` (no `Is` prefix). There's also `MPI_Finalized` that may be useful to protect the `MPI_Finalize` call.

amitmurthy commented:
I'll add a check to MPIManager so that a single instantiation of it can be used only once in an `addprocs` calls. You can have multiple different `manager` objects, so it will be possible to setup multiple MPI clusters.

In light of the above, I'll also add a check to `mpi_init` so that repeated invocations on the same object are harmless.

Checking via `MPI_Initialized` and `MPI_Finalized` should be done anyway - just to be resilient against the same calls made from other libraries.

andreasnoack commented:
Wouldn't it make sense to evaluate `@mpi_do` in `Main`? Therefore, it I load modules at the workers, it is difficult to use them.
```julia
julia> using MPI

julia> manager=MPIManager(np=4)
MPI.MPIManager(4,`mpirun -np 4 --output-filename /tmp/user/1021/juliaBBNbp0 julia --worker`,"/tmp/user/1021/juliaBBNbp0",60,Dict{Int64,Int64}(),Dict{Int64,Int64}(),PipeServer(active),"/tmp/user/1021/juliacv7Jfr")

julia> addprocs(manager)
4-element Array{Any,1}:
 2
 3
 4
 5

julia> @everywhere using MPI

julia> @everywhere using Base.LinAlg

julia> @everywhere LinAlg

julia> @mpi_do LinAlg
	From worker 3:	exception on 3: ERROR: LinAlg not defined
	From worker 3:	 in anonymous at multi.jl:1533
	From worker 3:	 in anonymous at multi.jl:844
	From worker 3:	 in run_work_thunk at multi.jl:603
	From worker 3:	 in run_work_thunk at multi.jl:612
	From worker 3:	 in anonymous at task.jl:6
	From worker 4:	exception on 4: ERROR: LinAlg not defined
	From worker 4:	 in anonymous at multi.jl:1533
	From worker 4:	 in anonymous at multi.jl:844
	From worker 4:	 in run_work_thunk at multi.jl:603
	From worker 4:	 in run_work_thunk at multi.jl:612
	From worker 4:	 in anonymous at task.jl:6
	From worker 2:	exception on 2: ERROR: LinAlg not defined
	From worker 5:	exception on 5: ERROR: LinAlg not defined
	From worker 2:	 in anonymous at multi.jl:1533
	From worker 5:	 in anonymous at multi.jl:1533
	From worker 2:	 in anonymous at multi.jl:844
	From worker 5:	 in anonymous at multi.jl:844
	From worker 2:	 in run_work_thunk at multi.jl:603
	From worker 5:	 in run_work_thunk at multi.jl:603
	From worker 2:	 in run_work_thunk at multi.jl:612
	From worker 5:	 in run_work_thunk at multi.jl:612
	From worker 2:	 in anonymous at task.jl:6
	From worker 5:	 in anonymous at task.jl:6
```
Right now it evaluates in `MPI`
```julia
julia> macroexpand(:(@mpi_do LinAlg))
quote  # /home/andreasnoackjensen/.julia/v0.4/MPI/src/cman.jl, line 71:
    begin  # task.jl, line 321:
        Base.sync_begin() # line 322:
        #45#v = begin  # /home/andreasnoackjensen/.julia/v0.4/MPI/src/cman.jl, line 72:
                for #44#p = MPI.workers() # line 73:
                    Base.spawnat(#44#p,$(Expr(:localize, :(()->begin  # expr.jl, line 110:
            ()->begin  # multi.jl, line 1533:
                    MPI.LinAlg
                end
        end))))
                end
            end # line 323:
        Base.sync_end() # line 324:
        #45#v
    end
end
```

andreasnoack commented:
Also, would it make sense to populate the `MPIManager` with the process numbers when calling `addprocs(MPIManager)` to keep track of them. Right now if I add socket workers after the MPI workers, `mpi_init` also registers ids of the socket workers
```julia
julia> using MPI

julia> manager=MPIManager(np=4)
MPI.MPIManager(4,`mpirun -np 4 --output-filename /tmp/user/1021/juliavdO9TD julia --worker`,"/tmp/user/1021/juliavdO9TD",60,Dict{Int64,Int64}(),Dict{Int64,Int64}(),PipeServer(active),"/tmp/user/1021/juliaxp0rTp")

julia> addprocs(manager)
4-element Array{Any,1}:
 2
 3
 4
 5

julia> addprocs(2)
2-element Array{Any,1}:
 6
 7

julia> @everywhere using MPI

julia> mpi_init(manager)

julia> manager.map_mpi_julia
Dict{Int64,Int64} with 8 entries:
  0 => 7
  4 => 2
  7 => 0
  2 => 4
  3 => 5
  5 => 3
  6 => 0
  1 => 3
```
Is that right?

andreasnoack commented on src/cman.jl:
```diff
@@ -0,0 +1,97 @@
+    for i in 1:nworkers()
+        mpi_id = fetch(refs[i][1])
+        manager.map_mpi_julia[mpi_id] = refs[i][2]
+        manager.map_mpi_julia[refs[i][2]] = mpi_id
```
should this be `manager.map_julia_mpi[refs[i][2]] = mpi_id`?

amitmurthy replied:
Yes.

amitmurthy commented:
Yes, to both executing `@mpi_do` under `Main` and having `mpi_init` only work off the workers it launches.

Will push an update.  Unfortunately due to #25 , I can only work on this from my system at the office and not on my laptop, so it sort of slows down things a bit.

amitmurthy commented:
Fixed as per discussion above.

`launch_cmd` is now `mpi_cmd` - it does not have the julia part
`@mpi_do` is now of the form `@mpi_do manager expr`
Removed the need for additional `using MPI` and `mpi_init`

A `manager` can be used only once:

```
julia> using MPI;manager=MPIManager(np=4); addprocs(manager)
4-element Array{Any,1}:
 2
 3
 4
 5

julia> addprocs(manager)
Reuse of an MPIManager is not allowed.
Try again with a different instance of MPIManager.
Error in MPI launch ErrorException("Reuse of MPIManager is not allowed.")
0-element Array{Any,1}

julia> manager2=MPIManager(np=4); addprocs(manager2)
4-element Array{Any,1}:
 6
 7
 8
 9

julia> @mpi_do manager println(myid())
        From worker 4:  4
        From worker 3:  3
        From worker 5:  5
        From worker 2:  2

julia> @mpi_do manager2 println(myid())
        From worker 9:  9
        From worker 6:  6
        From worker 7:  7
        From worker 8:  8

julia> @everywhere println(myid())
1
        From worker 2:  2
        From worker 4:  4
        From worker 3:  3
        From worker 5:  5
        From worker 6:  6
        From worker 7:  7
        From worker 9:  9
        From worker 8:  8
```

`procs(manager)` and `mpiprocs(manager)` have been added. See the README for the changed usage instructions.

Only openmpi is supported as of now.

lcw commented:
This looks fine to me but I am probably not the best to review this patch.  @andreasnoack Does the current update satisfy your concerns?  If so I would be happy to merge it in.

andreasnoack commented:
@amitmurthy Great changes. Thanks.

@lcw Yes. I think we should merge this. We can make changes afterwards if needed and it doesn't interfere with the original way of using this package.

eschnett commented:
Regarding redirecting `stdout` to a file: if MPICH doesn't support this during startup, then you can do this in C (or Julia). After calling `MPI_Init`, you would examine `argc` and `argv` to determine whether such a redirection is wanted. Then the code continues like this:
```
#include <stdio.h>
/* determine current process rank */
int rank;
MPI_Comm_rank(&rank, MPI_COMM_WORLD);
/* create a file name unique to this process */
char filename[1000];
snprintf(logfilename, sizeof filename, "stdout-%d.txt", rank);
/* redirect stdout */
printf("Redirecting stdout to %s -- no further output will appear here\n", logfilename);
fflush(stdout); /* not necessary */
FILE *logfile = freopen(logfilename, "w", stdout);
/* bonus points for checking whether FILE is NULL */
```
From there on, everything written to `stdout` from this process will instead appear in `logfile`. Redirecting `stderr` can be done in the same way.

I don't know how to access `stdout` from Julia, but apart from this, the call to `freopen` could easily appear in Julia code as well.

lcw commented:
Sounds good.  I will merge this for now, but we definitely want a MPI portable way of handling the output.  

I will wait to update METADATA.jl.  Does anybody know how to provide a different versions of MPI.jl to Julia depending on the julia version?

andreasnoack commented:
You closed the issue without merging!

lcw commented:
Thanks, should be merged now.  Sorry about that.

amitmurthy commented:
@eschnett the suggested scheme will only work for local processes. If the workers are on remote MPI nodes, either a) all the nodes must have access to a network file system or b) the MPI library must provide a mechanism to capture the stdout of the remote workers. The SSHManager does this by keeping the ssh session open after launch.

@lcw to address portability, I'll provide an optional scheme where initially the julia master listens on a known port and the workers send back their information to this port. 

Also, it is safe to update METADATA - since the REQUIRE file mentions `julia 0.4-`  for this commit, METADATA / Package manager will automatically make this update available only to julia 0.4 setups.

eschnett commented:
@amitmurthy Yes, I assumed that there is a single file system. If not, things get more complicated (but still possible):
- `stdout` of process 0 is very likely to be available to the caller of `mpirun`
- all other MPI processes could use `MPI_Send` to communicate with process 0, which then forwards this to the launcher, multiplexing over the same ssh channel (e.g. prefixing every line with the integer process rank)
- the launcher needs to de-multiplex the outputs
I'm not familiar with the details of the new SSHManager -- would this de-multiplexing work?

amitmurthy commented:
We don't use ssh here - I just mentioned SSHManager as an example of how it is done in the non-MPI case. I think a tcp listener on julia master, with all workers connecting to it and redirecting their STDOUTs would be the most portable way to go.

This, of course, will not work across public networks, say a laptop running master with the MPI workers on AWS, but we can design those type of use-cases separately.
